File: tutorial/java/Server/src/main/java/systems/zlink/tutorial/server/channel/RecordLoginHandler.java
package systems.zlink.tutorial.server.channel;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkSendHandler;
import systems.zlink.tutorial.shared.Contracts;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

// --8<-- [start:channel-send-handler]
// Handles a one-way message. There is no return value, so the caller is already
// done by the time this runs and cannot observe a failure here.
public final class RecordLoginHandler implements ZLinkSendHandler<Contracts.RecordLogin> {

    // The Framework owns the handler instance and builds it through Spring, so a
    // constructor parameter would have to be a bean. A logger is not one; the
    // ordinary Java idiom is a static logger per class.
    private static final Logger LOG = LoggerFactory.getLogger(RecordLoginHandler.class);

    @Override
    public CompletionStage<Void> handle(
            Contracts.RecordLogin message, ZLinkMessageContext context) {
        LOG.info("login recorded: {}", message.playerId());
        return CompletableFuture.completedFuture(null);
    }
}
// --8<-- [end:channel-send-handler]

File: tutorial/kotlin/Server/src/main/kotlin/systems/zlink/tutorial/server/channel/RecordLoginHandler.kt
package systems.zlink.tutorial.server.channel

import org.slf4j.LoggerFactory
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingSendHandler
import systems.zlink.tutorial.shared.RecordLogin

// --8<-- [start:channel-send-handler]
// Handles a one-way message. There is no return value, so the caller is already
// done by the time this runs and cannot observe a failure here.
@ZLinkHandlerGroup(HandlerGroups.PROFILE)
class RecordLoginHandler : ZLinkSuspendingSendHandler<RecordLogin> {

    private val log = LoggerFactory.getLogger(RecordLoginHandler::class.java)

    override suspend fun handle(message: RecordLogin, context: ZLinkMessageContext) {
        log.info("login recorded: {}", message.playerId)
    }
}
// --8<-- [end:channel-send-handler]
